### ISSUE TITLE
API Management (azurerm_api_management_api) support for openapi v3 content_format

### ISSUE THREAD

**Issue by mhennecke** [terraform-providers/terraform-provider-azurerm#3203]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The azurerm_api_management_api is missing support for OpenAPI v3 JSON and YAML imports in 
provider version v1.24.0. Please add "openapi", "openapi+json" and "openapi-link" to the allowed content formats.

### New or Affected Resource(s)

* azurerm_api_management_api 

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "test" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_api_management" "test" {
  name                = "example-apim"
  location            = "${azurerm_resource_group.test.location}"
  resource_group_name = "${azurerm_resource_group.test.name}"
  publisher_name      = "My Company"
  publisher_email     = "company@terraform.io"

  sku {
    name     = "Developer"
    capacity = 1
  }
}

resource "azurerm_api_management_api" "test" {
  name                = "example-api"
  resource_group_name = "${azurerm_resource_group.test.name}"
  api_management_name = "${azurerm_api_management.test.name}"
  revision            = "1"
  display_name        = "Example API"
  path                = "example"
  protocols           = ["https"]

  import {
    content_format = "openapi-link"
    content_value  = "https://raw.githubusercontent.com/OAI/OpenAPI-Specification/master/examples/v3.0/petstore.yaml"
  }
}
```

### References
* https://docs.microsoft.com/en-us/rest/api/apimanagement/apis/createorupdate#contentformat
* https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_api_management_api.go#L91

**Comment by tombuildsstuff** [comment]
hey @mhennecke 

Thanks for opening this issue :)

Taking a look into this it appears this requires an upgrade to the version of the Azure API we're using - [we've got an upstream issue open requesting that here](https://github.com/Azure/azure-rest-api-specs/issues/4409) (which I've just requested an update for), but once that becomes available in the Azure SDK for Go (which is generated from that repo) we can look to add support for this 👍

Thanks!

**Comment by tombuildsstuff** [comment]
👋 

The upstream issue's been fixed: https://github.com/Azure/azure-rest-api-specs/issues/4409 - once the SDK's been updated #3335 we should be able to go to the new version of the API Management API - at which point it should be possible to add support for this

Thanks!

**Comment by AndiA92** [comment]
hey @tombuildsstuff,

Any plans on this? :)

**Comment by AndiA92** [comment]
cc. @katbyte 
Do you have any plans on implementing this?

**Comment by ybelMekk** [comment]
Hei. Im intressed in this feature. Im working for Norway welfare and is in need for this, as we are exposing apis with the use of Azure Api-management. I would like to build the infrastructure with terraform.  

Any updates on the upgrade of the version of the Azure API?

please hit me back

thanks

**Comment by tombuildsstuff** [comment]
@ybelMekk @AndiA92 unfortunately we're still blocked on https://github.com/Azure/azure-rest-api-specs/issues/6372

**Comment by molinch** [comment]
@tombuildsstuff Hi Tom, is this issue blocking importing an OpenAPI specification using Terraform?
The above can be done using Powershell and ARM already.

**Comment by tombuildsstuff** [comment]
@molinch yep - we're still waiting on the API Team to publish a Swagger document (which in turn generates a Go SDK for us to use) for the new API version

**Comment by jeanpaulsmit** [comment]
Any updates on this or possibly a timeframe?

**Comment by timja** [comment]
Latest is:
https://github.com/Azure/azure-rest-api-specs/issues/6372#issuecomment-571308516

**Comment by PPACI** [comment]
Just for information. I've forked the provider on my side, and I've managed to used openapi3 json **and yaml**.
I think I will not do a PR since I don't have time to properly check inetgration test and if I didn't break something in the process, but here it what I've done :
1. I've updated the dependency of the resource to 2019-01-01 release of azure go sdk
https://github.com/terraform-providers/terraform-provider-azurerm/blob/a3fc3a99a8c8b5a412a906df662a54633d22f224/azurerm/internal/services/apimanagement/resource_arm_api_management_api.go#L9
I had to correct some signatures but it was more about passing empty string (that's where i'm not 100% sure of what I've done)
2. Add the new signature to the "enum" of valid content
https://github.com/terraform-providers/terraform-provider-azurerm/blob/a3fc3a99a8c8b5a412a906df662a54633d22f224/azurerm/internal/services/apimanagement/resource_arm_api_management_api.go#L102

For reference, here is the list
https://github.com/Azure/azure-sdk-for-go/blob/91316876b14205b98fc116bba77b3bb689746954/services/apimanagement/mgmt/2019-01-01/apimanagement/models.go#L205

Then you can `go build` and name your custom plugin like `terraform-provider-custom`. Put the provider in your project folder.

You will able to import it like
```hcl
provider "azurerm" {
  version = "=1.42.0"
}

provider "custom" {
}
```

and then use it **just for your resource** like
```hcl
resource "azurerm_api_management_api" "foobar" {
  provider            = custom
  name                 = "foobar"
  import {
    content_format = "openapi"
    content_value  = file("${path.module}/api-schema/openapi_openapi.yml")
  }
}
```

I hope it will help some people 👍
Since it's go, I can cross-compile the forked plugin for people if you don't want to re-fork it. Tell me.

PS. The fork is on an internal gitlab server, so I can't directly share the repo...

**Comment by mcalster** [comment]
@tombuildsstuff the Azure go sdk has been updated, do you have a timeline on when the Provider will be updated? Just wanting to know if we should pursue the steps @PPACI outlined above or wait for the official fix. 
Thx for the great work so far :-)

**Comment by timja** [comment]
It isn't from what I can tell?
https://github.com/Azure/azure-sdk-for-go/tree/master/services/apimanagement/mgmt

changes were in 2019-12-01-preview:
https://github.com/vfedonkin/azure-rest-api-specs/blob/87c99392a06f3fbfde0507d9c802c17aac48414a/specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimapis.json

**Comment by timja** [comment]
It has now been updated, released in 39.1.0 of the azure sdk for go https://github.com/Azure/azure-sdk-for-go/pull/7317

**Comment by mcalster** [comment]
Any progress on this?

**Comment by abhijithkumar123** [comment]
When I set content_format as "swagger-link-json" and try to import an openapi v3 spec file, I get the following error:

StatusCode=404 -- Original Error: autorest/azure: error response cannot be parsed: "" error: EOF

Is there a workaround for this?

**Comment by alanwales** [comment]
I am also seeing "The Swagger version specified is unknown" when trying to upload an open api v3 document, until the openapi-link content_format is supported we will continue to use V2 swagger

**Comment by ybelMekk** [comment]
Is this one closed any day soon? 💃

**Comment by vinoth12988** [comment]
@tombuildsstuff , Azure API stopped support on exporting Swagger-json format from the portal. Which means only way to update the existing API's in the APIM is through openAPI format from Terrafrom. This is now high priority.

**Comment by hctv19** [comment]
@tombuildsstuff There were a number of updates recently merged into the https://github.com/Azure/azure-rest-api-specs repo, and as I understand it this should allow the Go SDK to be regenerated  with the updated definitions needed to support OpenAPI content types in the Terraform provider.

**Comment by lucashuet93** [comment]
I'm getting started implementing the support for v3 content-formats now and will hopefully have a PR up by EOW.

**Comment by lucashuet93** [comment]
I'd began work on the addition of the OAS v3 content-formats, but realized that this is a much larger undertaking than originally expected. 

The current set of APIM related azurerm code uses the [2018-01-01](https://docs.microsoft.com/en-us/azure/templates/microsoft.apimanagement/2018-01-01/service/apis) version of APIM, and the OASv3 content-formats were added in version [2019-01-01](https://docs.microsoft.com/en-us/azure/templates/microsoft.apimanagement/2019-01-01/service/apis). The azurerm provider now uses version 41.0.0 of the Azure SDK for Go, which added support for APIM versions 2018-06-01-preview, 2019-01-01, and 2019-12-01-preview, so the addition of the OASv3 content-formats is indeed possible.

In order to make the change to allow the new content-formats on the API resource, however, the entire azurerm codebase for API Management needs to be updated to utilize 2019-01-01. As soon as the version is updated, it breaks clients and other resources downstream. The upgrade to the new version of APIM cannot be limited to the API resource type only. Even if it was possible to limit the changes to the API resource without causing other errors, I’m not sure that the Terraform team would accept a PR that causes the set of API Management resources to use multiple versions and be out of sync with each other.

**Comment by katbyte** [comment]
@lucashuet93 - i merged in a PR to upgrade to 2019-12-01 a couple days ago so this should be unblocked now 🙂

**Comment by lucashuet93** [comment]
@katbyte Awesome to hear! I've now opened PR #6618 which fixes this issue.

**Comment by mbfrahry** [comment]
This has been merged with #6618 and should make it into v2.8.0 of the provider
